crates/ruff_linter/src/rules/pylint/rules/unnecessary_chained_comparison.rs

```diff
@@ -0,0 +1,129 @@
+use std::collections::HashMap;
+
+use crate::checkers::ast::Checker;
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::{self as ast};
+
+/// ## What it does
+/// Checks for boolean operations such as `a < b and b < c`
+/// that can be refactored into a single comparison `a < b < c`.
+///
+/// ## Why is this bad?
+/// A single comparison is semantically clearer and more concise.
+///
+/// ## Example
+/// ```python
+/// a = int(input())
+/// b = int(input())
+/// c = int(input())
+/// if a < b and b < c:
+///     pass
+/// ```
+///
+/// Use instead:
+/// ```python
+/// a = int(input())
+/// b = int(input())
+/// c = int(input())
+/// if a < b < c:
+///     pass
+/// ```
+
+#[violation]
+pub struct UnnecessaryChainedComparison;
+
+impl Violation for UnnecessaryChainedComparison {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Simplified chain comparison exists between the operands.")
+    }
+}
+
+fn update_bounds<'a>(
+    operator: ast::CmpOp,
+    id: &'a str,
+    node_idx: u32,
+    is_left: bool,
+    uses: &mut HashMap<&'a str, (u32, u32)>,
+) {
+    // Store the bounds using bitwise operations. Each bit in these
+    // integers represents whether the identifier is involved in a lower or upper bound
+    // comparison at a specific node index.
+    //
+    // For example, if node_idx is 2, then:
+    // - `1 << node_idx` produces a bitmask with the 3rd bit set.
+    // - Using |=, we set the corresponding bit in the lower or upper bounds integer to 1 without modifying other bits.
+    //
+    // This efficiently tracks whether the identifier is used in lower or upper bounds
+    // comparisons across multiple nodes, allowing us to check for shared bounds later.
+    let entry = uses.entry(id).or_default();
+    match operator {
+        ast::CmpOp::Lt | ast::CmpOp::LtE if is_left => {
+            entry.0 |= 1 << node_idx; // Set the bit for lower bound
+        }
+        ast::CmpOp::Gt | ast::CmpOp::GtE if is_left => {
+            entry.1 |= 1 << node_idx; // Set the bit for upper bound
+        }
+        ast::CmpOp::Lt | ast::CmpOp::LtE if !is_left => {
+            entry.1 |= 1 << node_idx; // Set the bit for upper bound
+        }
+        ast::CmpOp::Gt | ast::CmpOp::GtE if !is_left => {
+            entry.0 |= 1 << node_idx; // Set the bit for lower bound
+        }
```

Comment:
Using bit-index here is clever, but I think it's a bit too clever and we can probably get a way with a simpler solution that also works when there are more than 32 nodes.

Correct me if I'm wrong, but I think the analysis has a few interesting properties

* The `node_idx` is always increasing (it's sorted!)
* The only data that we need to compute is the `shared_bounds`, `lower_bounds_count`, and `upper_bounds_count`

I wonder if we could use these properties to do the calculation right here in `update_bounds`:

What if we stored the following data for each identifier:

```rust
struct Bounds {
	lower_count: u32,
	upper_count: u32,
	shared_count: NonZeroU32,
	last_node_index: u32
}
```

* a lower bound increments `lower_count`
* an upper bound increments `upper_count`
* shared count gets updated when `last_node_index == node_idx`
* `update_bounds` always sets `last_node_index`.

Reply:
We just have to be careful with non-monotonic/non-totally-ordered comparisons (eg. `a < b > c and b < d`). The original pylint rule triggers on this case (although it's not clear what the correct fix is or if it is even appropriate to trigger on this case).